romancal/datamodels/container.py

```diff
@@ -0,0 +1,486 @@
+                try:
+                    init_from_asn = self.read_asn(init)
+                    self.from_asn(init_from_asn, asn_file_path=init)
+                except Exception as e:
+                    raise TypeError(
+                        "Input must be an ASN file or a list of either strings "
+                        "(full path to ASDF files) or Roman datamodels."
+                    ) from e
+            else:
+                raise TypeError(
+                    "Input must be an ASN file or a list of either strings "
+                    "(full path to ASDF files) or Roman datamodels."
+                )
+
+    def __len__(self):
+        return len(self._models)
+
+    def __getitem__(self, index):
+        if isinstance(index, slice):
+            start = index.start
+            stop = index.stop
+            step = index.step
+            m = self._models[start:stop:step]
+            m = [
+                rdm.open(item, memmap=self._memmap)
+                if (not isinstance(item, rdm.DataModel) and self._return_open)
+                else item
+                for item in m
+            ]
+        else:
+            m = self._models[index]
+            if not isinstance(m, rdm.DataModel) and self._return_open:
+                m = rdm.open(m, memmap=self._memmap)
+        return m
+
+    def __setitem__(self, index, model):
+        if isinstance(model, rdm.DataModel):
+            self._models[index] = model
+        else:
+            raise ValueError("Only datamodels can be used.")
+
+    def __delitem__(self, index):
+        del self._models[index]
+
+    def __iter__(self):
+        for model in self._models:
+            if not isinstance(model, rdm.DataModel) and self._return_open:
+                model = rdm.open(model, memmap=self._memmap)
+            yield model
+
+    def insert(self, index, model):
+        if isinstance(model, rdm.DataModel):
+            self._models.insert(index, model)
+        else:
+            raise ValueError("Only datamodels can be used.")
+
+    def append(self, model):
+        if isinstance(model, rdm.DataModel):
+            self._models.append(model)
+        else:
+            raise ValueError("Only datamodels can be used.")
+
+    def extend(self, input_object):
+        if not isinstance(input_object, (Iterable, rdm.DataModel)) or isinstance(
+            input_object, str
+        ):
+            raise ValueError("Not a valid input object.")
+        elif all(isinstance(x, rdm.DataModel) for x in input_object):
+            self._models.extend(input_object)
+        else:
+            raise ValueError("Not a valid input object.")
+
+    def pop(self, index=-1):
+        self._models.pop(index)
+
+    def copy(self, memo=None):
+        """
+        Returns a deep copy of the models in this model container.
+        """
+        return copy.deepcopy(self, memo=memo)
+
+    def close(self):
+        if not self._iscopy and self._asdf is not None:
```

Comment:
Does a Roman `ModelContainer` have a `_asdf` attribute? What is `self._asdf is not None` checking?